src/super_gradients/training/losses/iou_loss.py

```diff
@@ -0,0 +1,100 @@
+from typing import Union
+
+import torch
+
+from super_gradients.common.abstractions.abstract_logger import get_logger
+from super_gradients.training.losses.loss_utils import LossReduction
+from super_gradients.training.losses.structure_loss import AbstarctStructureLoss
+
+logger = get_logger(__name__)
+
+
+class IoULoss(AbstarctStructureLoss):
+    """
+    Compute average IoU loss between two tensors, It can support both multi-classes and binary tasks.
+    """
+    def _calc_numerator_denominator(self, labels_one_hot, predict):
+        """
+        Calculate iou metric's numerator and denominator.
+
+        :param labels_one_hot: target in one hot format.   shape: [BS, num_classes, img_width, img_height]
+        :param predict: predictions tensor.                shape: [BS, num_classes, img_width, img_height]
+        :return:
+            numerator = intersection between predictions and target.    shape: [BS, num_classes, img_width, img_height]
+            denominator = area of union between predictions and target. shape: [BS, num_classes, img_width, img_height]
+        """
+        numerator = labels_one_hot * predict
+        denominator = labels_one_hot + predict - numerator
+        return numerator, denominator
+
+    def _calc_loss(self, numerator, denominator):
+        """
+        Calculate iou loss.
+        All tensors are of shape [BS] if self.sum_over_batches else [num_classes]
+
+        :param numerator: intersection between predictions and target.
+        :param denominator: area of union between prediction pixels and target pixels.
+        """
+        loss = 1. - ((numerator + self.smooth) / (denominator + self.eps + self.smooth))
+        return loss
+
+
+class BinaryIoULoss(IoULoss):
+    """
+    Compute IoU Loss for binary class tasks (1 class only).
+    Except target to be a binary map with 0 and 1 values.
+    """
+    def __init__(self,
+                 apply_sigmoid: bool = True,
+                 smooth: float = 1.,
+                 eps: float = 1e-5):
+        """
+        :param apply_sigmoid: Whether to apply sigmoid to the predictions.
+        :param smooth: laplace smoothing, also known as additive smoothing. The larger smooth value is, closer the IoU
+            coefficient is to 1, which can be used as a regularization effect.
+            As mentioned in: https://github.com/pytorch/pytorch/issues/1249#issuecomment-337999895
+        :param eps: epsilon value to avoid inf.
+        """
+        super().__init__(apply_softmax=False, ignore_index=None, smooth=smooth, eps=eps, sum_over_batches=False)
+        self.apply_sigmoid = apply_sigmoid
+
+    def forward(self, predict, target):
+        if self.apply_sigmoid:
+            predict = torch.sigmoid(predict)
+        return super().forward(predict=predict, target=target)
+
+
+class GeneralizedIoULoss(IoULoss):
```

Comment:
you'll only want to use this class for class imbalance, otherwise use regular `IoULoss`.

Reply:
You can always use the base `IoULoss`, this subclass is more kind of a guide of how to properly use generalized loss.
Since both nominator and denominator are normalized by `1 / sqrt(counts)` and `counts` is positive large number, i.e num of Tp pixels in the minibatch, the denominator turns into a small number. Therefore `eps` must be much smaller, and smooth is removed as suggested in the generalized class,  to prevent skewed results.